collector/metrics_collector.go

```diff
@@ -0,0 +1,154 @@
+package collector
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type MetricsCollector struct {
+	clusterHealth *ClusterHealth
+	nodes         *Nodes
+	indices       *Indices
+
+	clusterHealthResponse *clusterHealthResponse
+	nodeStatsResponse     *nodeStatsResponse
+	indexStatsResponse    *indexStatsResponse
+}
+
+func NewMetricsCollector(logger log.Logger, client *http.Client, url *url.URL, all bool, exportIndices bool) *MetricsCollector {
+	return &MetricsCollector{
+		clusterHealth: NewClusterHealth(logger, client, url),
+		nodes:         NewNodes(logger, client, url, all),
+		indices:       NewIndices(logger, client, url, all, exportIndices),
+	}
+}
+
+func (c *MetricsCollector) Describe(ch chan<- *prometheus.Desc) {
+	c.clusterHealth.Describe(ch)
+	c.nodes.Describe(ch)
+	c.indices.Describe(ch)
+}
+
+func (c *MetricsCollector) Collect(ch chan<- prometheus.Metric) {
+	// clusterHealth
+	clusterHealthResponse, err := c.clusterHealth.fetchAndDecodeClusterHealth()
+	if err != nil {
+		c.clusterHealth.up.Set(0)
+		level.Warn(c.clusterHealth.logger).Log(
+			"msg", "failed to fetch and decode cluster health",
+			"err", err,
+		)
+		return
+	}
+	c.clusterHealth.up.Set(1)
+
+	// nodes
+	nodeStatsResponse, err := c.nodes.fetchAndDecodeNodeStats()
+	if err != nil {
+		c.nodes.up.Set(0)
+		level.Warn(c.nodes.logger).Log(
+			"msg", "failed to fetch and decode node stats",
+			"err", err,
+		)
+		return
+	}
+	c.nodes.up.Set(1)
+
+	// indices
+	indexStatsResponse, err := c.indices.fetchAndDecodeIndexStats()
+	if err != nil {
+		c.indices.up.Set(0)
+		level.Warn(c.indices.logger).Log(
+			"msg", "failed to fetch and decode index stats",
+			"err", err,
+		)
+		return
+	}
+	c.indices.up.Set(1)
+
+	c.clusterHealth.Collect(ch, clusterHealthResponse)
+	c.nodes.Collect(ch, nodeStatsResponse)
+	c.indices.Collect(ch, clusterHealthResponse, nodeStatsResponse, indexStatsResponse)
```

Comment:
wrapping collectors (or your custom non-interface-fulfilling Collect() ) isn't a good idea. Doing so makes the collectors run sequentially, not in parallel. Also the json retrieval works parallel, if you exclude it from the collectors. 
In additionMaking collectors optional is then just a matter of `if`ing the `prometheus.Mustregister()`.

Could you please refer to you initial intent of this PR: get the docs metrics of the primary shards. To archieve this I would strongly suggest to
 - go back to the start (sorry)
 - implement the indices collector (fulfilling the collector interface) with just the very few metrics included you are interested in. We would really like to keep the overall index metrics we can get via `_nodes/stats` in the nodes collector. Please only export the structs marshaled from the `_all/_stats` that are not in `_nodes/stats`.
 - add this collector optionally with `prometheus.MustRegister`